index.bs

```diff
@@ -990,6 +990,174 @@ while (true) {
+
+<pre class="example" highlight="js">
+// Adds an entry to the event log on the page, optionally applying a specified
+// CSS class.
+function addToEventLog(text, severity = 'info') {
+  let log = document.getElementById('event-log');
+  let mostRecentEntry = log.lastElementChild;
+  let entry = document.createElement('li');
+  entry.innerText = text;
+  entry.className = 'log-' + severity;
+  log.appendChild(entry);
+
+  // If the most recent entry in the log was visible, scroll the log to the
+  // newly added element.
+  if (mostRecentEntry != null &&
+      mostRecentEntry.getBoundingClientRect().top <
+          log.getBoundingClientRect().bottom) {
+    entry.scrollIntoView();
+  }
+}
+
+// "Connect" button handler.
+async function connect() {
+  let url = document.getElementById('url').value;
+  try {
+    var transport = new WebTransport(url);
+    addToEventLog('Initiating connection...');
+  } catch (e) {
+    addToEventLog('Failed to create connection object. ' + e, 'error');
+    return;
+  }
+
+  try {
+    await transport.ready;
+    addToEventLog('Connection ready.');
+  } catch (e) {
+    addToEventLog('Connection failed. ' + e, 'error');
+    return;
+  }
+
+  transport.closed
+      .then(() => {
+        addToEventLog('Connection closed normally.');
+      })
+      .catch(() => {
+        addToEventLog('Connection closed abruptly.', 'error');
+      });
+
+  globalThis.currentTransport = transport;
+  globalThis.streamNumber = 1;
+  try {
+    globalThis.currentTransportDatagramWriter = transport.datagramWritable.getWriter();
+    addToEventLog('Datagram writer ready.');
+  } catch (e) {
+    addToEventLog('Datagram writer failed. ' + e, 'error');
+    return;
+  }
+  readDatagrams(transport);
+  acceptUnidirectionalStreams(transport);
+  document.forms.sending.elements.send.disabled = false;
+  document.getElementById('connect').disabled = true;
+}
+
+// "Send data" button handler.
+async function sendData() {
+  let form = document.forms.sending.elements;
+  let encoder = new TextEncoder('utf-8');
+  let rawData = sending.data.value;
+  let data = encoder.encode(rawData);
+  let transport = globalThis.currentTransport;
+  try {
+    switch (form.sendtype.value) {
+      case 'datagram':
+        await globalThis.currentTransportDatagramWriter.write(data);
+        addToEventLog('Sent datagram: ' + rawData);
+        break;
+      case 'unidi': {
+        let stream = await transport.createUnidirectionalStream();
+        let writer = stream.writable.getWriter();
+        await writer.write(data);
+        writer.close();
+        addToEventLog('Sent a unidirectional stream with data: ' + rawData);
+        break;
+      }
+      case 'bidi': {
+        let stream = await transport.createBidirectionalStream();
+        let number = globalThis.streamNumber++;
+        readFromIncomingStream(stream, number);
+
+        let writer = stream.writable.getWriter();
+        await writer.write(data);
+        writer.close();
```

Comment:
I think since bidirectional streams don't support a half-closed state, this will prevent reading data sent by the server. Probably better to just not close here, and rely on the server to close the stream.

Reply:
Opened Issue https://github.com/w3c/webtransport/issues/163 to clarify.

Reply:
Looks like bi-directional streams are intended to support a half-closed state (though a bug in the current implementation doesn't allow this). The example runs anyway, because the server opens a separate bi-directional stream on which it returns the count.